### DISCUSSION TITLE
How to type the return value of `getStaticProps` in Next.js with Prisma's generated types?

### DISCUSSION THREAD

**Discussion by marceloverdijk** [prisma/prisma#4372] [Q&A] [answered]
I'm experimenting with Next.js and Prisma and I wonder how I can do the below best.
I'm having a `Circuit` model which has a relation to a `Country` which I want to include in the query and pass on in the React components.

It looks like:

```ts
import { Circuit, PrismaClient } from '@prisma/client'
import { GetStaticPaths, GetStaticProps } from 'next'
import CircuitListDetail from '../../components/circuit/CircuitListDetail'
import Layout from '../../components/Layout'

type Props = {
  item?: Circuit
}

const StaticPropsDetail = ({ item }: Props) => {
  return (
    <Layout
      title={`${
        item ? item.name : 'Circuit Detail'
      } | Next.js + Prisma + TypeScript Example`}
    >
      {item && <CircuitListDetail item={item} />}
    </Layout>
  )
}

export default StaticPropsDetail

export const getStaticPaths: GetStaticPaths = async () => {
  const prisma = new PrismaClient()
  const circuits: Circuit[] = await prisma.circuit.findMany()
  return {
    paths: circuits.map((circuit) => ({
      params: {
        id: circuit.id
      }
    })),
    fallback: false
  }
}

export const getStaticProps: GetStaticProps = async ({ params }) => {
  const prisma = new PrismaClient()
  const item = await prisma.circuit.findUnique({
    where: {
      id: params?.id as string
    },
    include: { 
      country: true 
    }
  })
  return {
    props: {
      item
    }
  }
}
```

In the `getStaticProps` function I've added `include: { country: true }` to also query the country.
But the `Props` contains a `item?: Circuit` which does not have the country...

I was reading about the `PromiseReturnType` in https://www.prisma.io/docs/concepts/components/prisma-client/advanced-usage-of-generated-types but honestly I don't know where to start... 

And maybe to make it more complicated, the `Country` model also has a `Continent` relation I would also like to include when querying the `Circuit`.

**Comment by nikolasburk** [comment, marked as answer]
Hey Marcel, that's a great question! You're on the right path here, but in this case you need to use the generated `CircuitGetPayload` type rather than `PromiseReturnType`.

If I understand correctly, this would probably solve your issue:

```ts
import { Prisma } from '@prisma/client'

type CircuitWithCountry = Prisma.CircuitGetPayload<{
  include: { country: true }
}>

type Props = {
  item?: CircuitWithCountry
}

export const getStaticProps: GetStaticProps<Props> = async ({ params }) => {
  const prisma = new PrismaClient()
  const item = await prisma.circuit.findUnique({
    where: {
      id: params?.id as string
    },
    include: { 
      country: true 
    }
  })
  return {
    props: {
      item
    }
  }
}
```

> Note that the way how `CircuitGetPayload` is exported has changed in last Tuesday's release [2.12.0](https://github.com/prisma/prisma/releases/tag/2.12.0). It's now accessible via the exported `Prisma` namespace. If you're using an older version, you can still export it directly: `import { CircuitGetPayload } from '@prisma/client'`

Let me know if that works of if you have any further questions!

**Reply by marceloverdijk** [reply]
Thx for your quick feedback, I will test it this evening and will let you know.

**Reply by marceloverdijk** [reply]
Yes it worked!! Thx again. Prisma is a really powerful combination with Next.js to build a SSG.

For other readers, I had to to use:

`export const getStaticProps: GetStaticProps = async ({ params }) => {`

instead of:

`export const getStaticProps: GetStaticProps<Props> = async ({ params }) => {`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
